src/ieee.c

```diff
@@ -113,41 +558,180 @@ create_directory_listing(uint8_t *data)
 			}
 		}
 		*data++ = '"';
-		if (namlen > 16) {
-			namlen = 16; // TODO hack
-		}
+		//if (namlen > 16) {
+		//	namlen = 16; // TODO hack
+		//}
 		memcpy(data, dp->d_name, namlen);
 		data += namlen;
 		*data++ = '"';
 		for (int i = namlen; i < 16; i++) {
 			*data++ = ' ';
 		}
 		*data++ = ' ';
-		*data++ = 'P';
+		if (S_ISDIR(st.st_mode)) {
+			*data++ = 'D';
+			*data++ = 'I';
+			*data++ = 'R';
+		} else {
+			*data++ = 'P';
+			*data++ = 'R';
+			*data++ = 'G';
+		}
+		// This would be a '<' if file were protected, but it's a space instead
+		*data++ = ' ';
+
+		if (dirlist_timestmaps) {
+			// ISO-8601 date+time
+			struct tm mtime;
+			if (localtime_r(&st.st_mtime, &mtime)) {
+				*data++ = ' '; // space before the date
+				data += strftime((char *)data, 20, "%Y-%m-%d %H:%M:%S", &mtime);
+			}
+		}
+		
+		*data++ = 0;
+		return data - data_start;
+	}
+
+	// link
+	*data++ = 1;
+	*data++ = 1;
+
+	*data++ = 255; // "65535"
+	*data++ = 255;
+
+	memcpy(data, blocks_free, strlen(blocks_free));
+	data += strlen(blocks_free);
+	*data++ = 0;
+
+	// link
+	*data++ = 0;
+	*data++ = 0;
+	(void)closedir(dirlist_dirp);
+	dirlist_eof = true;
+	return data - data_start;
+}
+
+
+static int
+create_cwd_listing(uint8_t *data)
+{
+	uint8_t *data_start = data;
+	int file_size;
+
+	// load address
+	*data++ = 1;
+	*data++ = 8;
+	// link
+	*data++ = 1;
+	*data++ = 1;
+	// line number
+	*data++ = 0;
+	*data++ = 0;
+	*data++ = 0x12; // REVERSE ON
+	*data++ = '"';
+	for (int i = 0; i < 16; i++) {
+		*data++ = ' ';
+	}
+	if (cgetcwd((char *)data - 16, 16)) {
+		dirlist_eof = true;
+		return 0;
+	}
+	*data++ = '"';
+	*data++ = ' ';
+	*data++ = 'H';
+	*data++ = 'O';
+	*data++ = 'S';
+	*data++ = 'T';
+	*data++ = ' ';
+	*data++ = 0;
+
+	char *tmp = malloc(strlen(hostfscwd)+1);
+	if (tmp == NULL) {
+		set_error(0x70, 0, 0);
+		return 0;
+	}
+	int i = strlen(hostfscwd);
+	int j = strlen(fsroot_path);
+	strcpy(tmp,hostfscwd);
+
+	for(;;) {
```

Comment:
This `for(;;)` "feels" out-of-place to me, mostly because I tend to get bothered by for-loops that are programmed like `while(true)`-loops, and it seems like it would be appropriate to loop on `for(; i >= j-1; --i)`. 

In particular, the fact that this is programmed like an infinite loop leaves me worried about line 673, below, which is `if (!namlen) continue;`, which appears to be an infinite-loop case unless `tmp[i-1]=0;` on line 664 is making a change that would cause us to decrement and continue via the `if` on line 659.

tl;dr: Is line 673 a bug? And if so, would it be more appropriate to make the stopping condition and iteration behavior of the for-loop be more explicit by placing them in the `for()` statement rather than manually decrementing `i` and checking the break condition at the end of the loop?

Reply:
Here's the logical flow of that entire for loop with example strings:

`hostfscwd = "//home/whatever"; fsroot_path = "/";`
- i = 15
- j = 1
- first 8 iterations of the for loop, the first if statement matches, `i` gets decremented and the loop restarts
- the 9th iteration, the if statement is false and it continues on
- tmp, which started out as `"//home/whatever"` is now `"//home\x00whatever"`
- i = 7 now, and is not < j
- file_size = 0 ; the file size string output is copied from the regular directory listing code, so we kept this variable even though it's constant.
- namlen is now 8 as tmp+i is "whatever", and is not zero
- we add the path component `"whatever"` to the output of the directory listing and decrement i, continuing the loop
- starting with i=6, we hit true on the first if statement 4 times
- i now = 2, and the if condition is false because tmp[1] is a /
- we make that spot null, so now tmp = `"/\x00home\x00whatever"`
- i is still < j
- namlen is 4
- output `"home"` as path component and decrement i
- starting with i=1 the if is immediately false
- tmp becomes `"\x00\x00home\x00whatever"`
- namlen is now 0, and we repeat the loop without decrement
- this time the first if is true because we clobbered the leading `/`
- i is now 0 and is less than j so tmp becomes `"/"`
- and this is the final path element that gets appended to the data output

tl;dr, namlen is never 0 unless tmp[i-1] is not a path separator character.  And the loop will not get stuck because either the first if will be true, or the second one, if (i < j), is true.

Reply:
Come to think of it, all of the `continue`s are either associated with an `i--` or for the `namlen` check, an `i--` would simply save one iteration and not change the important part of the flow.  We can put `i--` in the third arg of `for()`.  As for the check condition, we can also get rid of the `break` with your suggestion.